### PR TITLE
Issue #8: Add support for composer merge plugin

### LINKED ISSUE #8: Does not integrate well with wikimedia/composer-merge-plugin

### PR DESCRIPTION
**Pull request by fabianx-ai**
This adds direct support for the composer-merge-plugin by wikimedia and has no more bad hacks.

If the proper version of wikimedia/merge-plugin is installed it will just work, else nothing changes.

This allows a structure to work like:

composer.base.json: Define all the common variables.
composer.core.json: Include composer.base.json, override commit-prefix, commit-branch-prefix
composer.drupal.json: Include composer.core.json, which means patches no longer need to be specified in both locations.

Fixes #8.
### Notes

Depends on composer-1.1, which was just recently released, so would be a v2.4.0.

### CODE REVIEW

File: src/Drupal/Composer/DrupalInstallerMergePluginTrait.php [hunk 1]
```diff
@@ -0,0 +1,40 @@
+<?php
+
+namespace Drupal\Composer;
+
+use Composer\Script\Event;
+use Composer\Script\ScriptEvents;
+
+trait DrupalInstallerMergePluginTrait {
```
Comment (douggreen) [inline src/Drupal/Composer/DrupalInstallerMergePluginTrait.php]: Why is this a trait? Is there a missing file that also uses it?

Comment (fabianx-ai) [inline src/Drupal/Composer/DrupalInstallerMergePluginTrait.php]: The other file has been added now.

**Comment by douggreen** [comment]
Sorry, I thought I'd commented on this. When I tested using merge plugin, the settings merged in such a way that made it unusable. I imagined that the settings in the merged composer.json would apply to the require's in that composer.json, and that the settings in my local composer.json applied to the requires in my local one. But that doesn't seem to be how this works. Comments?

**Comment by fabianx-ai** [comment]
Unfortunately this only works properly with https://github.com/wikimedia/composer-merge-plugin/pull/114 applied. dev-master#[commit-hash-of-PR]

I had hoped it would have made more progress till now.

Note:

The behavior can be changed with replace: true / false,

replace: true:

``` php
$rootArray = drupal_array_merge_deep($rootArray, $includedArray);
```

replace: false:

``` php
$rootArray = drupal_array_merge_deep($includedArray, $rootArray);
```

So my use case was to use this not for local overrides, but for creating a base config and to use a chain like:

composer.base.json: All settings
composer.core.json: include base and overwrite just git.commit-prefix and git.commit-message (or what that is called).
composer.drupal.json: include core.json (including all the core patches, so they don't need to be duplicated)
composer.libraries.json: include composer.base.json

which all would use replace: false.

Unfortunately it is not possible to mix replace: true with replace: false, so at this moment, you can either have local overrides OR basing configuration in an inheritance like structure.

Both are principally supported by this PR though.

Once composer-1.1 is stable, we can use the INIT event instead though, so this has just some temporary merit.

**Comment by fabianx-ai** [comment]
@douggreen This is ready again!

**Comment by douggreen** [comment]
Thanks!

**Pull request merged**